rl/approximate_dynamic_programming.py

```diff
@@ -167,17 +169,22 @@ def backward_evaluate(
 
     '''
     v: List[FunctionApprox[S]] = []
+    
+    num_steps: int = len(mrp_f0_mu_triples)
 
     for i, (mrp, approx0, mu) in enumerate(reversed(mrp_f0_mu_triples)):
 
         def return_(s_r: Tuple[S, float], i=i) -> float:
-            s, r = s_r
-            return r + γ * (v[i-1].evaluate([s]).item() if i > 0 else 0.)
+            s1, r = s_r
+            return r + γ * (v[i-1].evaluate([s1]).item() if i > 0 and not
+                            mrp_f0_mu_triples[num_steps - i][0].is_terminal(s1)
+                            else 0.)
 
         v.append(
             approx0.solve(
                 [(s, mrp.transition_reward(s).expectation(return_))
-                 for s in mu.sample_n(num_state_samples)],
+                 for s in mu.sample_n(num_state_samples)
+                 if not mrp.is_terminal(s)],
```

Comment:
With this approach, you might get a lot less than `n` states sampled here if there are a lot of terminal states. Could this be a problem?

Reply:
Yes, but I'm not sure there's a way around it if we are sampling states with a simulation trace. If we have an analytic form for the non-terminal states distribution (in some examples, we do), then we can sample directly from that known distribution of non-terminal states (for a given time step).

Reply:
Okay, that makes sense. Sounds like there's no point worrying about it unless it really becomes a problem.